### ISSUE TITLE
`no-top-level-side-effects` disallows `delete`

### ISSUE THREAD

**Issue by kevinoid** [sindresorhus/eslint-plugin-unicorn#3633]
The [no-top-level-side-effects](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-top-level-side-effects.md) rule disallows `delete` statements, while allowing assignments.  These seem to have similar potential for "side-effects" to me.  I wonder if this may be an oversight?

For example:

```js
const report = process.report.getReport();
report.sharedObjects = undefined;  // ok
delete report.sharedObjects;  // error
export default report;
```

Obviously the issue can be worked around by using destructuring in most cases, but it would be nice to avoid unnecessary changes without good reason.

Thanks for considering, and for this useful new rule!
Kevin

**Comment by sindresorhus** [comment]
This is intentional. Assignments are a narrow usability exception, not a general exemption for mutations. `delete` is observably different from assigning `undefined`, and allowing it would also permit clear import-time effects such as `delete globalThis.foo`. Distinguishing harmless local deletions would require flow analysis beyond this intentionally conservative rule. For shaping an exported object, please use object rest destructuring or a local rule disable.

**Comment by kevinoid** [comment]
Makes sense to me.  Thanks for explaining the rationale, and especially for updating the docs for others.  I appreciate it!
